post.md

```diff
@@ -0,0 +1,411 @@
+on the other hand first method may cause stack overflow in some cases.
+
+This is just one of many examples where low level details (assembly, network
+stack, devices, etc.) condition higher level code execution. That's why any
+senior software developer should know them. And of course any security
+specialist. Except that it's essentional to anyone interested in Reverse
+Enineering.
+
+This article is meant to preasent the most important ideas behind x86 assembly
+just to show how does it work, what are its limitations. If you want to code
+in assembly or read disassembly I recommend to look at [x86 instruction set]
+(https://c9x.me/x86/) and tutorials [like this](https://www.nayuki.io/page/a-fundamental-introduction-to-x86-assembly-programming).
+If you are interested in advanced optimization you'd have to dive into CPU
+model specific documentation.
+
+Basics of x86 CPU
+-----------------
+
+We have CPU — computation unit and RAM for temporary data
+storage. We are usually separated from all other devices but
+we use OS interface, which reuses concepts we use for
+those two. Generally speaking, everything that happens in 
+a computer is series of passing data between components and
+transforming them in between.
+
+For example, when we display a web page we can think of it:
+
+1. we have given an address in memory.
+2. transform it into request conforming standard.
+3. pass a request to the networking device
+4. accept response
+5. transform text into an image
+6. pass the image to graphics card
+
+The job of application is just to transform the data and
+inform OS where is the product, what it is and where to pass
+it. RAM is a medium for that exchange as well as storage for
+middle products. RAM also store instructions how to perform
+those transformations. Assembly language is a textual
+representation of codes understood by CPU.
+
+# Memory
+
+We can think of RAM as a function. Every byte in memory has
+its ordinal number. We use it to read or change its value.
+Bytes are usually accessed in groups of 4(32bit) or 8(64-bit).
+
+X86 CPU never use two memory locations at the same time. That's
+why it has own memory called registers. In modern x86
+architecture, there are 16 64-bit general purpose registers
+called: RAX, RBX, RCX, RDX, RDI (destination index), RSI
+(source index), RBP (base pointer), RSP (stack pointer),
+R8-R15. Despite meaningful names of some, only RSP preserved special
+meaning (it was important when it was usual to code directly
+in assembly). "R" letter in beginning denote its 64-bits as
+during 40 years of x86 evolution registers grown from 16-bit
+(AX, BX, etc), to 32-bit (EAX, EBX, etc). Among special
+registers, we have RIP (instruction pointer) stores pointer
+(= ordinal number in memory) of next instruction to run and
+EFLAGS which register special situations (like zeroing
+register or overflow while adding). Those special registers
+are never accessed directly.
```

Comment:
X86 CPU never uses two memory locations at the same time. That's why it has own memory called registers. In modern x86 architecture, there are 16 64-bit general purpose registers called: RAX, RBX, RCX, RDX, RDI (destination index), RSI (source index), RBP (base pointer), RSP (stack pointer), R8-R15. Despite meaningful names of some, only RSP preserved special meaning (it was important when it was usual to code directly in assembly). The letter "R" at the beginning of denote its 64-bits as during 40 years of x86 evolution registers grown from 16-bit (AX, BX, etc), to 32-bit (EAX, EBX, etc). Among special registers, we have RIP (instruction pointer) stores pointer (= ordinal number in memory) of the next instruction to run and EFLAGS which registers special situations (like zeroing register or overflow while adding). Those special registers are never accessed directly.